### PR TITLE
[Hotfix][PLAT-992] Keep download count for files over multiple nodes

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Currently when a file moves to a different node the download count is "lost", but really it's just always filtering by it's target node/preprint/whatever. This fix ensures we only filter by the file id. 

## Changes

- creates a new classmethod for page counter which filters only by action and file id
- removes now superfluous utility function `get_basic_counters`

## QA Notes

You should be able to move files from node to node and retain download counts. Also it's important to remember this works retroactively, so old files will now have old download counts from different nodes correctly added in. 

## Documentation

🐞  fix no docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/PLAT-992

### CODE REVIEW

File: osf/models/files.py [hunk 1]
```diff
@@ -331,25 +330,15 @@ def touch(self, auth_header, revision=None, **kwargs):
...
-            parts.append(version)
-        page = ':'.join([format(part) for part in parts])
-        _, count = get_basic_counters(page)
-
-        return count or 0
-
     def get_download_count(self, version=None):
         """Pull the download count from the pagecounter collection"""
-        return self.get_page_counter_count('download', version=version)
+        from osf.models import PageCounter
```
Comment (Johnetordoff) [inline osf/models/files.py]: Circular imports prevented by doing this here and in `get_view_count`

File: osf/models/files.py [hunk 2]
```diff
@@ -331,25 +330,15 @@ def touch(self, auth_header, revision=None, **kwargs):
...
+        return PageCounter.get_file_downloads(self._id, version)
 
     def get_view_count(self, version=None):
         """Pull the mfr view count from the pagecounter collection"""
-        return self.get_page_counter_count('view', version=version)
+        from osf.models import PageCounter
+        return PageCounter.get_file_views(self._id)
```
Comment (caseyrollins) [inline osf/models/files.py]: drive-by: should `version` be passed here?

File: osf/models/analytics.py [hunk 3]
```diff
@@ -136,9 +136,27 @@ def update_counter(cls, page, node_info):
...
+        :param version: Int the version number
+        :return: Int the total count
+        """
+        if version is not None:
+            file_id += ':' + str(version)
+
         try:
-            counter = cls.objects.get(_id=cls.clean_page(page))
-            return (counter.unique, counter.total)
+            return PageCounter.objects.filter(_id__startswith=counter_type,
```
Comment (caseyrollins) [inline osf/models/analytics.py]: Why is it necessary to aggregate the total? `PageCounter` `_id`'s are unique so wouldn't this be something like: 
```
counter = PageCounter.objects.get(_id__startswith=counter_type, _id__endswith=file_id)
return (counter.unique, counter.total)
```

Comment (caseyrollins) [reply osf/models/analytics.py]: Oh wait... nevermind. There might be several records with a certain file id in the `_id` but with different node ids.

File: osf/models/files.py (hunk 2 above)
Comment (Johnetordoff) [reply osf/models/files.py]: 👍

File: osf/models/analytics.py [hunk 4]
```diff
@@ -136,9 +136,27 @@ def update_counter(cls, page, node_info):
             model_instance.save()
 
     @classmethod
-    def get_basic_counters(cls, page):
+    def _get_file_counter(cls, counter_type, file_id, version=None):
+        """
+        Pulls the total file count for views or downloads across different nodes.
+        :param info_type: String
```
Comment (erinspace) [inline osf/models/analytics.py]: Change this to `counter_type`

File: osf/models/analytics.py [hunk 5]
```diff
@@ -136,9 +136,27 @@ def update_counter(cls, page, node_info):
...
+        if version is not None:
+            file_id += ':' + str(version)
+
         try:
-            counter = cls.objects.get(_id=cls.clean_page(page))
-            return (counter.unique, counter.total)
+            return PageCounter.objects.filter(_id__startswith=counter_type,
+                                              _id__endswith=':' + file_id).aggregate(total=models.Sum('total'))['total'] or 0
         except cls.DoesNotExist:
-            return (None, None)
```
Comment (erinspace) [inline osf/models/analytics.py]: Since you've changed this query from a `.get()` to a `.filter()`, you should no longer need this `try` `except` block. In fact, pretty sure you don't need the `or 0` part at the end of your aggregate query, as it should return 0 if nothing is found regardless.

File: osf/models/files.py (hunk 1 above)
Comment (erinspace) [reply osf/models/files.py]: The circular import can also be avoided by importing as `from osf.models.analytics import PageCounter` at the top

File: osf/models/files.py [hunk 6]
```diff
@@ -331,25 +330,15 @@ def touch(self, auth_header, revision=None, **kwargs):
...
     def get_download_count(self, version=None):
         """Pull the download count from the pagecounter collection"""
-        return self.get_page_counter_count('download', version=version)
+        from osf.models import PageCounter
+        return PageCounter.get_file_downloads(self._id, version)
 
     def get_view_count(self, version=None):
         """Pull the mfr view count from the pagecounter collection"""
-        return self.get_page_counter_count('view', version=version)
+        from osf.models import PageCounter
```
Comment (erinspace) [inline osf/models/files.py]: Ditto!

File: osf/models/analytics.py (hunk 5 above)
Comment (Johnetordoff) [reply osf/models/analytics.py]: Aggregating across nothing returns a NoneType not zero, so that is necessary, but the try-catch is not.

Comment (erinspace) [reply osf/models/analytics.py]: Strange, when I tried it locally seemed to return 0, but maybe I was doing something different. No biggie either way!

Comment (Johnetordoff) [reply osf/models/analytics.py]: Yeah really what's happening is that files with zero views/downloads don't have Pagecounters at all so they will return none, but if you create a Pagecounter object with zero views/downloads and sum over it it will actually return zero, but that shouldn't happen in a real environment as for as I know.

File: osf/models/analytics.py [hunk 7]
```diff
@@ -158,9 +158,24 @@ def update_counter(cls, page, node_info):
...
+    def _get_file_counter(cls, counter_type, file_id, version=None):
+        """
+        Pulls the total file count for views or downloads across different nodes.
+        :param counter_type: String
+        :param file_id: String
+        :param version: Int the version number
+        :return: Int the total count
+        """
+        if version is not None:
+            file_id += ':' + str(version)
```
Comment (sloria) [inline osf/models/analytics.py]: It's generally a bad idea to "mutate" function arguments if you can avoid it (strings are immutable so you're not actually mutating file_id, but the concept still holds). The caller would not expect a "getter" function to have side effects.

Also, use `str.format`, as per style guide: https://cosdev.readthedocs.io/en/latest/style_guides/python.html#string-formatting .
